### PR TITLE
switch from nose to pytest (and pytest-cov), see also #70

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
pytest is a more advanced testing framework than nose (IMHO) and makes writing tests a
better experience as one has to type less and has more powerful mechanisms at hand.

Also, in case of a test failure, the output is more helpful in finding the cause.

See: http://pytest.org/latest/

With this changeset, the amount of tests and test fails (none for normal tests, some complaints for pylint) has not changed.

Currently, pytest runs in "compatibility mode", so it is happily executing your existing "written-for-nose" tests.

But future tests (and also existing tests) can now be written with less and prettier code.

### CODE REVIEW

**Comment by jdkasten** [comment]
I don't have a strong opinion on python testing frameworks.  I will checkout pytest tomorrow afternoon and the proposed changes.

Assuming there isn't any discussion and everything looks good.  I will merge this tomorrow night.

Thanks for PR!

File: setup.py [hunk 1]
```diff
@@ -20,9 +20,8 @@
 ]
 
 testing_extras = [
-    'coverage',
-    'nose',
-    'nosexcover',
+    'pytest',
+    'pytest-cov',
```
Comment (kuba) [inline setup.py]: `pytest-cov` is not available e.g. in Debian stable (https://tracker.debian.org/pkg/python-pytest-cov) and hence causes addiitonal packaging problems on top of #139... Therefore, I would keep nosetest for now.

File: tox.ini [hunk 2]
```diff
@@ -4,17 +4,12 @@
...
 [testenv]
 commands =
     python setup.py dev
-    python setup.py test -q # -q does not suppress errors
-
-[testenv:cover]
-commands =
-    python setup.py dev
-    python setup.py nosetests --with-coverage --cover-min-percentage=47
+    py.test --cov letsencrypt
```
Comment (kuba) [inline tox.ini]: Does `py.test` have corresponding `cover-min-percentage`?

File: tox.ini [hunk 3]
```diff
@@ -4,17 +4,12 @@
 # and then run "tox" from this directory.
 
 [tox]
-envlist = py26,py27,cover,lint
+envlist = py26,py27,lint
```
Comment (kuba) [inline tox.ini]: I believe separating out `cover` is really beneficial as it shows us what is wrong with the code: whether it breaks existing tests or fails to maintain minimum coverage percentage level.

File: setup.py (hunk 1 above)
Comment (ThomasWaldmann) [inline setup.py]: a) testing tools are usually just needed for developers and not installed for users of a software.
developers use virtualenv and pip, so they don't need the debian stable package, even if they use debian stable.
b) until letsencrypt is finished, the current debian testing might become the new stable and it has pytest-cov. testing is already frozen IIRC.

File: tox.ini (hunk 2 above)
Comment (ThomasWaldmann) [inline tox.ini]: I did only a quick search for it, but didn't find it.
Guess it would be rather pytest-cov that would have something like that.

File: tox.ini (hunk 3 above)
Comment (ThomasWaldmann) [inline tox.ini]: The reasons why I put cover into same run as normal testing are:
a) less waiting for tests with coverage to complete, guess it is about -25% if it is done within a normal test run and not needed a whole separate test run just for coverage
b) it might get executed more often if it is part of normal test run and people do not need to invoke it separately. so people might actually look at it more often and have less test commands to invoke (if not using tox)
c) as it is now, it is just showing covery, so no breaking anything AFAIK

File: setup.py (hunk 1 above)
Comment (kuba) [inline setup.py]: It's worth noting that `nose` is used solely as the test runner (discovery).

It is true that the testing tools are for devs, but please remember that `letsencrypt.test` module is intentionally distributed as a part of the core source tree in order to allow any user to run such tests, etc. (background reading for such design decision can be easily googled). On the other hand, I guess that you're suggesting `py.test` not only to allow `assert x == y`-like tests syntax, but also to use `import pytest` and use `pytest.*` functions inside test modules, am I right? If so, then `pytest` would need to be included inside `install_requires` just like in case of `mock` (as opposed to `testing_extras`). Luckily, `pytest` itself seems to be available in Debian stable, and it's rather unnecessary to include problematic `pytest-cov` in `install_requires`.

@jdkasten has a fair aim to support as many operating systems and environments as possible, hence e.g. support for Python 2.6. While you're correct about jessie freeze, note that even squeeze is still alive and marked as LTS! My comment about Debian stable was just an example, but it seems to me that `nose` is much more popular and it's worth checking the support in other OSes before proceeding.

File: tox.ini (hunk 3 above)
Comment (kuba) [inline tox.ini]: I was afraid that "it is just showing coverage" vs. failing if coverage is below the required threshold (100% in our case, although temporarily lowered by @jdkasten to have more green in Travis CI :)) and thus by including this commit we will no longer track (through Travis) coverage properly: users will be able to submit PRs without tests that will show up as green :(

Comment (kuba) [inline tox.ini]: `pytest --cov-min` is available only in yet unreleased 2.0 branch: https://github.com/schlamar/pytest-cov/commit/8f679db735ab99f66808e79d4e401996a9191ec0

File: tox.ini (hunk 2 above)
Comment (kuba) [inline tox.ini]: `pytest --cov-min` is available only in yet unreleased 2.0 branch: https://github.com/schlamar/pytest-cov/commit/8f679db735ab99f66808e79d4e401996a9191ec0

**Comment by jdkasten** [comment]
Sorry about the delay with my comments on this PR.  It does seem that Pytest has nice facilities.  As @kuba mentioned, my main concern would be future compatibility issues.  We will need to be able to support as many systems as possible.  I know there is demand to have the client run on atypical and older systems.

(@kuba I know I am far behind... but I am determined to monotonically increase the code coverage requirement.)

**Comment by ThomasWaldmann** [comment]
well, you may have a conflict of interests/goals here.

on the one hand, you want devs happily / comfortably writing lots of (good) tests, on the other hand you want to make legacy system users happy AND even be able to run your tests on the legacy system AND satisfy any testing dependency from legacy system packages.

I know this kind of conflict (had it myself for MoinMoin), so just some more infos to help you decide:
- users who run a legacy system (python < 2.7, maybe in general weak python support) maybe won't jump on brand new software like letsencrypt in the near future anyway.
- if they do, they maybe didn't do it the first time and have a more recent python installed (self-compiled or 3rd party repo) and are maybe used to assembling stuff as they need it (potentially also creating backports of packages as needed).
- as time goes by, these problems become less important as people upgrade to newer systems, but other problems become more important (like good test coverage, comfortable test framework, porting to python3)

About pytest:

- a lot of projects use this (since many years), likely everybody who ever tried it practically. so while it may be new to you, it's not new in general.
- the community around it is rather active
- if you'ld like to be able to use older packages you can restrict yourself to a feature set present in some older pytest version and specify that as minimum requirement.

Other options:
- installing the whole testsuite and test framework for end users is not quite the usual thing, but I can follow you that for security critical software it's maybe a nice thing to do. if for some legacy system this is causing troubles, whoever packages it for these could still drop the testing stuff from the package in case the legacy system can not support it. Or bundle whatever is needed.

To summarize: at some time you've to decide what's more important for you(comfortable devs or legacy users, the future or the past). BTW, this will come up again with py3 porting. To do this with reasonable effort and with keeping the src clean and readable, you'll likely have to drop 2.6 and py3 <= 3.2.

**Comment by jdkasten** [comment]
Thanks @ThomasWaldmann for the comments and direction.  I have been trying to elicit further comments/opinions about what needs to be supported.  I would like to establish a plan so that we can be consistent with decisions going forward.  (especially before we decide to apply any one-way functions to the codebase) I am sorry for stalling a bit on this topic.

**Comment by ThomasWaldmann** [comment]
note: travis ci build fail is due to old linter issue (it always failed before @kuba cleaned up).

**Comment by kuba** [comment]
It seems like the current configuration works out quite well in practice and doesn't have the aforementioned compatibility issues. Do you think this PR can be closed for now, @ThomasWaldmann?

**Comment by ThomasWaldmann** [comment]
well, above I already layed out all facts and pros and cons I see. it's your choice what tools you use, but don't complain if people who are used to better tools are unhappy with them (and it's their choice then how much they want to contribute in such circumstances).
